### PR TITLE
Mods

### LINKED ISSUE #95: Mods are not being installed on initial boot
Right now if someone configures his server with mods and run the server using `dst-server start` the mods are never installed. (Use `--update=mods` to workaround this problem).

And if they run `dst-server update --mods` before they run a `dst-server start` the mods are not updated/installed because the file `dedicated_server_mods_setup.lua` is not generated.

### PR DESCRIPTION
**Pull request by BraisGabin**
Fixes #95 

I added bats (IT IS GREAT!!!) to test the new features. I'll move all our tests to bats later (we don't need bats to launch `0.8`.

### CODE REVIEW

File: build/Dockerfile [hunk 1]
```diff
@@ -1,6 +1,16 @@
 FROM dstacademy/steamcmd:0.3
 MAINTAINER DST Academy <support@dst.academy>
 
+# Install dependencies.
```
Comment (thasmo) [inline build/Dockerfile:4]: For what reasons has this been moved atop?

File: build/Dockerfile [hunk 2]
```diff
@@ -63,6 +63,18 @@ RUN set -x \
 	&& mkdir -p $(dirname $CLUSTER_PATH) \
 	&& chown $STEAM_USER:$STEAM_USER $(dirname $CLUSTER_PATH)
 
+ARG MODS
+ENV INSTALLED_MODS ${MODS}
```
Comment (thasmo) [inline build/Dockerfile]: Can we name the `ENV` variable also `MODS` instead of `INSTALLED_MODS`?

File: .travis.yml [hunk 3]
```diff
@@ -10,12 +10,14 @@ env:
...
   - sudo add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable"
+  - sudo add-apt-repository ppa:duggan/bats --yes
   - sudo apt-get update
-  - sudo apt-get -y install docker-ce
+  - sudo apt-get -y install docker-ce bats
 
 script:
   - docker build -t $DOCKER_IMAGE_NAME ./build
   - ./test/run.sh
+  - bats tests
```
Comment (thasmo) [inline .travis.yml]: I'd rather not add some new tests using `bats` in this PR but add them after the `0.8` release. I know this is not a problem in general, but it keeps the codebase clean and consistent after all. Could you port the `bats` tests into our current test-suite, without using `bats` in this PR?

File: docs/configuration.md [hunk 4]
```diff
@@ -25,6 +25,13 @@ NAME=Name
 MAX_PLAYERS=10
 ```
 
+## Build arguments
```
Comment (thasmo) [inline docs/configuration.md]: I'd prefer `Build Arguments`; using an uppercase `A`.

File: build/Dockerfile [hunk 5]
```diff
@@ -63,6 +63,18 @@ RUN set -x \
...
+
+# Install mods
```
Comment (thasmo) [inline build/Dockerfile]: `Install mods.`; add a dot `.`. Sorry for nitpicking. :D

File: build/Dockerfile [hunk 6]
```diff
@@ -63,6 +63,18 @@ RUN set -x \
 	&& mkdir -p $(dirname $CLUSTER_PATH) \
 	&& chown $STEAM_USER:$STEAM_USER $(dirname $CLUSTER_PATH)
 
+ARG MODS
```
Comment (thasmo) [inline build/Dockerfile]: The `ARG/ENV` lines should be moved to the existing `ARG/ENV` block.

File: build/Dockerfile [hunk 7]
```diff
@@ -63,6 +63,18 @@ RUN set -x \
...
+ENV INSTALLED_MODS ${MODS}
+
+# Install mods
+RUN set -x \
```
Comment (thasmo) [inline build/Dockerfile]: I'd prefer moving the `Install mods` block directly below the `Install Don't Starve Together Server` block, if possible.

File: docs/configuration.md [hunk 8]
```diff
@@ -25,6 +25,13 @@ NAME=Name
...
+
+**MODS**
```
Comment (thasmo) [inline docs/configuration.md]: There need to be two spaces after `**MODS**` to force the line-break in Markdown.

File: docs/configuration.md [hunk 9]
```diff
@@ -25,6 +25,13 @@ NAME=Name
...
+Defines mods to install and enable.
+- *CSV of workshop IDs*
```
Comment (thasmo) [inline docs/configuration.md]: Also two spaces here, on the end of the line, please. 👍

**Review by thasmo — COMMENTED** [review]
Thanks for your awesome work! Please have a look at my feedback and feel free to comment.

File: .travis.yml (hunk 3 above)
Comment (BraisGabin) [reply .travis.yml]: I can remove all the tests from bats and add them later in other PR.

File: build/Dockerfile (hunk 1 above)
Comment (BraisGabin) [reply build/Dockerfile:4]: If you change an argument you don't want to install the dependencies again. So they have to be atop.

File: build/Dockerfile (hunk 6 above)
Comment (BraisGabin) [reply build/Dockerfile]: Same reason. I don't want to download the game again only for a change in the mods.

File: build/Dockerfile (hunk 2 above)
Comment (BraisGabin) [reply build/Dockerfile]: I prefer to keep `INSTALLED_MODS` as a _private_ `ENV`. So a different name is more _secure_.

File: .travis.yml (hunk 3 above)
Comment (thasmo) [reply .travis.yml]: Thanks! :)

**Comment by BraisGabin** [comment]
Updated!

File: build/Dockerfile (hunk 2 above)
Comment (thasmo) [reply build/Dockerfile]: I'm not sure I can follow. How does it differ by using `MODS` only?

Comment (BraisGabin) [reply build/Dockerfile]: Hmmm... I was thinking about this, and right now, split it have no sense. My idea in this PR was to remove the `--update=mods` option. But then I re-thought it. This change was made with that in mind.

**Comment by BraisGabin** [comment]
Updated again removing the `INSTALLED_MODS` change.

**Comment by BraisGabin** [comment]
I just changed a bit my mind. With this changes this PR doesn't break the retro-compatibility.

So this PR just allows you to add the mods in the image but you can install them the old way too.

Anyway, the documentation only points to the new and more _dockerish_ way.

**Comment by thasmo** [comment]
I understand that rebuilding images, after changing build arguments, is costly when moving the argument definitions above the `RUN` commands, because potential cache layers are lost. Nevertheless I think we still should move the `ARG` and `ENV` definitions to the top, because that way the Dockerfile is easier to read and understand in general, which is a topic overlooked too often.

Would you mind moving them to the top again, please? After that I think we're ready to merge this one. Thanks a lot!

**Comment by BraisGabin** [comment]
Do you mean to move the ARG definition above the Install dependencies or move the MODS up to the same block of the rest ARGs? or both?

Keep all the ARGs together... ok, I that is easier to read. But there's no need to be the first block in the Dockerfile.

I think that when a server admin is preparing his server s/he is going to try with lots of mod convinations. Download the game every single time s/he does a test is too much.

And even further, if you mean both s/he is going to dowload the dependencies again too. And they _never_ change.

**Review by thasmo — APPROVED** [review]

**Pull request merged**